### PR TITLE
Add extra tests and time filter

### PR DESCRIPTION
**Pull request by znicholls**

### CODE REVIEW

**Comment by danielhuppmann** [comment]
Sorry for the delay, @znicholls, had my hands full with diapers...

The changes look good to me, and you're the man on the time formatting, so I assume that this works as you'd like it to.

Not sure I understand your question on the timeseries, though - what do you mean by "accessing meta"? The reason why I set `iamc_index=True` by default (i.e., ignore `extra_cols`) is to not change behaviour in any of the IPCC notebooks.

File: pyam/core.py [hunk 1]
```diff
@@ -1132,7 +1136,45 @@ def _apply_filters(data, meta, filters):
             keep_col = pattern_match(data[col], values, level, regexp)
 
         elif col == 'year':
-            keep_col = years_match(data[col], values)
+            try:
+                keep_col = years_match(data[col], values)
+            except KeyError as exc:
+                if str(exc) != "'year'":
+                    raise exc
+                elif 'time' in data.columns:
```
Comment (danielhuppmann) [inline pyam/core.py]: Better to use `if self.time_col is not ‘time’`

File: pyam/core.py [hunk 2]
```diff
@@ -1132,7 +1136,45 @@ def _apply_filters(data, meta, filters):
...
+                    keep_col = years_match(data['time'].apply(lambda x: x.year),
+                                           values)
+                else:
+                    _raise_filter_error(col)
+
+        elif col == 'month':
+            if 'time' not in data.columns:
```
Comment (danielhuppmann) [inline pyam/core.py]: See above

File: pyam/core.py [hunk 3]
```diff
@@ -1132,7 +1136,45 @@ def _apply_filters(data, meta, filters):
...
+                _raise_filter_error(col)
+            else:
+                keep_col = month_match(data['time'].apply(lambda x: x.month),
+                                       values)
+
+        elif col == 'day':
+            if 'time' not in data.columns:
```
Comment (danielhuppmann) [inline pyam/core.py]: See above

File: pyam/core.py [hunk 4]
```diff
@@ -1132,7 +1136,45 @@ def _apply_filters(data, meta, filters):
...
+                wday = (isinstance(values, str)
+                        or (isinstance(values, list) and isinstance(values[0], str)))
+                if wday:
+                    days = data['time'].apply(lambda x: x.weekday())
+                else:  # ints or list of ints
+                    days = data['time'].apply(lambda x: x.day)
+                keep_col = day_match(days, values)
+
+        elif col == 'hour':
+            if 'time' not in data.columns:
```
Comment (danielhuppmann) [inline pyam/core.py]: See above

File: pyam/core.py (hunk 1 above)
Comment (znicholls) [reply pyam/core.py]: that would be great except we don't have access to `self` in `_apply_filters`, thoughts?

Comment (danielhuppmann) [reply pyam/core.py]: right...

would it make sense to pass `self.time_col` as a required arg through `_apply_filters`, or is that making it needlessly complicated?

Comment (znicholls) [reply pyam/core.py]: I'm a big fan of trying to make these internal conventions as clear as possible. Having said that, I think my refactoring would be to actually move `_apply_filters` so it was a method of `IamDataFrame`. The reason being that `_apply_filters` is specific enough in what it expects (in particular the pattern matching) that it only makes sense to apply it to `IamDataFrame` instances, not a general dataframe. So I think options are:

1. leave it
1. pass time_col as required arg
1. refactor and move `_apply_filters` so it's a method of `IamDataFrame` (my preference but also the most complicated)

Comment (znicholls) [reply pyam/core.py]: ok I'll do it that way then (when I get around to it..)

**Comment by znicholls** [comment]
@danielhuppmann getting this up to speed now. I'll leave the test, `tests/test_core.py::test_init_df_with_extra_col` failing as that requires your update of the behaviour of timeseries. Otherwise it should soon be good to go.

**Comment by znicholls** [comment]
@danielhuppmann tidying this up now. I'll leave the failing test as that requires your change to the behaviour of timeseries and meta. Everything else should be good to go soon.

**Comment by danielhuppmann** [comment]
thanks @znicholls for the new filter features and the refactoring!

**Pull request merged**